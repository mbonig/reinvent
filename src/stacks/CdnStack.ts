import { Stack, StackProps } from 'aws-cdk-lib';
import { ICertificate } from 'aws-cdk-lib/aws-certificatemanager';
import {
  Distribution,
  OriginProtocolPolicy,
  OriginRequestPolicy,
  ViewerProtocolPolicy,
} from 'aws-cdk-lib/aws-cloudfront';
import { LoadBalancerV2Origin } from 'aws-cdk-lib/aws-cloudfront-origins';
import { ILoadBalancerV2 } from 'aws-cdk-lib/aws-elasticloadbalancingv2';
import { Construct } from 'constructs';

interface CDNStackProps extends StackProps {
  websiteDomainName: string;
  certificate: ICertificate;
  websiteLoadBalancer: ILoadBalancerV2;
}

export class CdnStack extends Stack {
  constructor(scope: Construct, id: string, props: CDNStackProps) {
    super(scope, id, props);

    const fqdn = `${props.websiteDomainName}.matthewbonig.com`;
    const originRequestPolicy = new OriginRequestPolicy(this, 'OriginRequestPolicy', {
      headerBehavior: {
        behavior: 'allViewer',
      },
    });
    new Distribution(this, 'Resource', {
      defaultBehavior: {
        origin: new LoadBalancerV2Origin(props.websiteLoadBalancer, {
          protocolPolicy: OriginProtocolPolicy.HTTPS_ONLY,
        }),
        originRequestPolicy: {
          originRequestPolicyId: originRequestPolicy.originRequestPolicyId,
        },
        viewerProtocolPolicy: ViewerProtocolPolicy.REDIRECT_TO_HTTPS,
      },
      certificate: props.certificate,
      domainNames: [fqdn],
    });
  }

}
