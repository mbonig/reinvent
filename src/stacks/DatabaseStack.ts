import { RemovalPolicy, Stack, StackProps } from 'aws-cdk-lib';
import { AttributeType, ITableV2, StreamViewType, TableV2 } from 'aws-cdk-lib/aws-dynamodb';
import { Construct } from 'constructs';

export interface DatabaseStackProps extends StackProps {
  tablePrefix: string;
}

export class DatabaseStack extends Stack {
  public tables: ITableV2[] = [];

  constructor(scope: Construct, id: string, props: DatabaseStackProps) {
    super(scope, id, props);

    new TableV2(this, 'HeroesTable', {
      tableName: `${props.tablePrefix}-Heroes`,
      partitionKey: {
        name: 'PK',
        type: AttributeType.STRING,
      },
      sortKey: {
        name: 'SK',
        type: AttributeType.STRING,
      },
      dynamoStream: StreamViewType.NEW_AND_OLD_IMAGES,
      removalPolicy: RemovalPolicy.RETAIN_ON_UPDATE_OR_DELETE,
      replicas: [
        {
          region: 'us-west-2',
        },
      ],
    });

  }

}
