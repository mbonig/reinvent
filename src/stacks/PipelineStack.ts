import { Stack, StackProps } from 'aws-cdk-lib';
import { Secret } from 'aws-cdk-lib/aws-secretsmanager';
import { CodePipeline, CodePipelineSource, DockerCredential, ShellStep } from 'aws-cdk-lib/pipelines';
import { Construct } from 'constructs';
import { DEV_ACCOUNT, PRIMARY_REGION, PROD_ACCOUNT } from '../constants';
import { ApplicationStage } from '../stages/ApplicationStage';

interface PipelineStackProps extends StackProps {
  replicationRoleArn: string;
  connectionArn: string;
}

export type WebsiteProps = {
  websiteDomainName: string;
  replicationRoleArn: string;
}

export class PipelineStack extends Stack {
  constructor(scope: Construct, id: string, props: PipelineStackProps) {
    super(scope, id, props);

    const devWebsiteProps: WebsiteProps = {
      websiteDomainName: 'reinvent-dev',
      replicationRoleArn: props.replicationRoleArn,
    };
    const prodWebsiteProps: WebsiteProps = {
      websiteDomainName: 'reinvent',
      replicationRoleArn: props.replicationRoleArn,
    };

    const pipeline = new CodePipeline(this, 'Pipeline', {
      crossAccountKeys: true,
      synth: new ShellStep('Synth', {
        input: CodePipelineSource.connection('mbonig/reinvent', 'main', {
          connectionArn: props.connectionArn,
        }),
        commands: [
          'yarn',
          'yarn build',
          'npx cdk synth',
        ],
      }),
      dockerCredentials: [
        DockerCredential.dockerHub(new Secret(this, 'DockerHub', {})),
      ],
    });

    this.createDevWave(pipeline, devWebsiteProps);
    this.createProdWave(pipeline, prodWebsiteProps);
  }

  private createProdWave(pipeline: CodePipeline, prodWebsiteProps: WebsiteProps) {
    const prodWave = pipeline.addWave('Prod', {});

    const prodPrimaryStage = new ApplicationStage(this, 'ProdPrimary', {
      env: {
        account: PROD_ACCOUNT,
        region: PRIMARY_REGION,
      },
      envTag: 'prod',
      ...prodWebsiteProps,
      vpc: {
        cidr: '10.2.0.0/24',
      },
      tablePrefix: 'prod',
    });

    prodWave.addStage(prodPrimaryStage, {
      stackSteps: prodPrimaryStage.stackSteps,
    });
  }

  private createDevWave(pipeline: CodePipeline, devWebsiteProps: WebsiteProps) {
    const devWave = pipeline.addWave('Dev', {});

    const devPrimaryStage = new ApplicationStage(
      this,
      'DevPrimary',
      {
        env: {
          account: DEV_ACCOUNT,
          region: PRIMARY_REGION,
        },
        envTag: 'dev',
        ...devWebsiteProps,
        vpc: {
          cidr: '10.1.0.0/24',
        },
        tablePrefix: 'dev',
      });

    devWave.addStage(devPrimaryStage);

  }
}
