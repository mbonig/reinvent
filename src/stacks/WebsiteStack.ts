import { Duration, Stack, StackProps, Tags } from 'aws-cdk-lib';
import { Certificate, CertificateValidation, ICertificate } from 'aws-cdk-lib/aws-certificatemanager';
import { ITableV2 } from 'aws-cdk-lib/aws-dynamodb';
import { IVpc } from 'aws-cdk-lib/aws-ec2';
import { Platform } from 'aws-cdk-lib/aws-ecr-assets';
import { Cluster, ContainerImage } from 'aws-cdk-lib/aws-ecs';
import { ApplicationLoadBalancedFargateService } from 'aws-cdk-lib/aws-ecs-patterns';
import { ILoadBalancerV2 } from 'aws-cdk-lib/aws-elasticloadbalancingv2';
import { HostedZone } from 'aws-cdk-lib/aws-route53';
import { IBucket } from 'aws-cdk-lib/aws-s3';
import { Construct } from 'constructs';
import { ReplicatedBucket } from '../constructs/ReplicatedBucket';
import { WebsiteStackTest } from '../constructs/WebsiteStackTest/WebsiteStackTest';

export interface WebsiteStackProps extends StackProps {
  tablePrefix: string;
  replicationRoleArn: string;
  websiteDomainName: string;
  tables: ITableV2[];
  vpc: IVpc;
}


export class WebsiteStack extends Stack {
  mediaBucket: IBucket;
  loadBalancer: ILoadBalancerV2;
  certificate: ICertificate;

  constructor(scope: Construct, id: string, props: WebsiteStackProps) {
    super(scope, id, props);
    const { vpc, tables } = props;
    const hostedZone = HostedZone.fromLookup(this, 'HostedZone', { domainName: 'matthewbonig.com' });
    const cluster = new Cluster(this, 'Cluster', {
      vpc,
    });
    const mediaBucket = this.mediaBucket = new ReplicatedBucket(this, 'MediaBucket', {
      websiteDomainName: props.websiteDomainName,
      region: this.region,
    });

    const regionSpecificDomainName = `${props.websiteDomainName}-${Stack.of(this).region}.matthewbonig.com`;

    let certificate = this.certificate = new Certificate(this, 'Certificate', {
      domainName: `${props.websiteDomainName}.matthewbonig.com`,
      subjectAlternativeNames: [regionSpecificDomainName],
      validation: CertificateValidation.fromDns(hostedZone),
    });
    const service = new ApplicationLoadBalancedFargateService(this, 'Service', {
      cluster,
      domainName: regionSpecificDomainName,
      domainZone: hostedZone,
      certificate: certificate,
      redirectHTTP: true,
      healthCheckGracePeriod: Duration.minutes(3),
      desiredCount: 2,
      taskImageOptions: {
        image: ContainerImage.fromAsset('website', {
          platform: Platform.custom('linux/amd64'),
        }),
        containerPort: 3000,
        environment: {
          MEDIA_BUCKET: mediaBucket.bucketName,
        },
      },
    });

    this.loadBalancer = service.loadBalancer;

    mediaBucket.grantWrite(service.taskDefinition.taskRole, 'uploads/*');
    mediaBucket.grantRead(service.taskDefinition.taskRole, 'media/*');

    for (const table of tables) {
      table.grantReadWriteData(service.taskDefinition.taskRole);
    }

    const tests = new WebsiteStackTest(this, 'UrlPingTest', {
      url: `https:${regionSpecificDomainName}`,
    });
    tests.node.addDependency(service);

    Tags.of(this).add('role', 'website');
  }

}
