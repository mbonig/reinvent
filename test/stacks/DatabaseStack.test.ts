import { App } from 'aws-cdk-lib';
import { Template } from 'aws-cdk-lib/assertions';
import { DatabaseStack, DatabaseStackProps } from '../../src/stacks/DatabaseStack';

describe('DatabaseStack Snapshot', () => {
  function getTestStack(props?: Partial<DatabaseStackProps>) {
    const app = new App();
    const stack = new DatabaseStack(app, 'test', {
      tablePrefix: 'test',
      env: {
        account: '000011112222',
        region: 'us-east-1',
      },
      ...props,
    });

    return Template.fromStack(stack);
  }

  test('matches snapshot', () => {
    const assert = getTestStack();
    expect(assert.toJSON()).toMatchSnapshot();
  });
});
