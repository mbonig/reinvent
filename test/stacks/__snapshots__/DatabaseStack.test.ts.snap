// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DatabaseStack Snapshot matches snapshot 1`] = `
{
  "Parameters": {
    "BootstrapVersion": {
      "Default": "/cdk-bootstrap/hnb659fds/version",
      "Description": "Version of the CDK Bootstrap resources in this environment, automatically retrieved from SSM Parameter Store. [cdk:skip]",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": {
    "HeroesTable555F4C83": {
      "DeletionPolicy": "RetainExceptOnCreate",
      "Properties": {
        "AttributeDefinitions": [
          {
            "AttributeName": "PK",
            "AttributeType": "S",
          },
          {
            "AttributeName": "SK",
            "AttributeType": "S",
          },
        ],
        "BillingMode": "PAY_PER_REQUEST",
        "KeySchema": [
          {
            "AttributeName": "PK",
            "KeyType": "HASH",
          },
          {
            "AttributeName": "SK",
            "KeyType": "RANGE",
          },
        ],
        "Replicas": [
          {
            "Region": "us-west-2",
          },
          {
            "Region": "us-east-1",
          },
        ],
        "StreamSpecification": {
          "StreamViewType": "NEW_AND_OLD_IMAGES",
        },
        "TableName": "test-Heroes",
      },
      "Type": "AWS::DynamoDB::GlobalTable",
      "UpdateReplacePolicy": "Retain",
    },
  },
  "Rules": {
    "CheckBootstrapVersion": {
      "Assertions": [
        {
          "Assert": {
            "Fn::Not": [
              {
                "Fn::Contains": [
                  [
                    "1",
                    "2",
                    "3",
                    "4",
                    "5",
                  ],
                  {
                    "Ref": "BootstrapVersion",
                  },
                ],
              },
            ],
          },
          "AssertDescription": "CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI.",
        },
      ],
    },
  },
}
`;
